src/main/java/com/es/phoneshop/dao/CustomProductDao.java

```diff
@@ -0,0 +1,126 @@
+        result.add(new Product("iphone6", "Apple iPhone 6", new BigDecimal(1000), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Apple/Apple%20iPhone%206.jpg"));
+        result.add(new Product("htces4g", "HTC EVO Shift 4G", new BigDecimal(320), usd, 3, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/HTC/HTC%20EVO%20Shift%204G.jpg"));
+        result.add(new Product("sec901", "Sony Ericsson C901", new BigDecimal(420), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Sony/Sony%20Ericsson%20C901.jpg"));
+        result.add(new Product("xperiaxz", "Sony Xperia XZ", new BigDecimal(120), usd, 100, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Sony/Sony%20Xperia%20XZ.jpg"));
+        result.add(new Product("nokia3310", "Nokia 3310", new BigDecimal(70), usd, 100, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Nokia/Nokia%203310.jpg"));
+        result.add(new Product("palmp", "Palm Pixi", new BigDecimal(170), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Palm/Palm%20Pixi.jpg"));
+        result.add(new Product("simc56", "Siemens C56", new BigDecimal(70), usd, 20, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C56.jpg"));
+        result.add(new Product("simc61", "Siemens C61", new BigDecimal(80), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C61.jpg"));
+        result.add(new Product("simsxg75", "Siemens SXG75", new BigDecimal(150), usd, 40, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg"));
+        return result;
+    }
+
+    @Override
+    public Optional<Product> getProduct(Long id) {
+        if (id != null) {
+            readLock.lock();
+            try {
+                return productList.stream()
+                        .filter(product -> id.equals(product.getId()))
+                        .findAny();
+            } finally {
+                readLock.unlock();
+            }
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public List<Product> findProducts() {
+        readLock.lock();
+        try {
+            return productList.stream()
+                    .filter(product -> product.getPrice() != null && product.getStock() > 0)
+                    .toList();
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public void save(Product product) {
+        if (product == null) {
+            throw new NullPointerException("Product was null");
+        }
+        writeLock.lock();
+        try {
+            Optional<Product> productWithSameId = getProduct(product.getId());
```

Comment:
can't a deadlock happen here?
and why did you change the implementation of this method via productList.contains(product)? It was fine in while the previous review
if you want to do an update for product here, then navigating by stream is better idea than calling getProduct() inside the write lock